src/OmniSharp.Stdio/StdioResponseStream.cs

```diff
@@ -0,0 +1,187 @@
+using System;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Newtonsoft.Json;
+using OmniSharp.Stdio.Protocol;
+using OmniSharp.Stdio.Services;
+
+namespace OmniSharp.Stdio
+{
+    public class StdioResponseStream : Stream
+    {
+        private readonly ISharedTextWriter _sharedWriter;
+        private readonly ResponsePacket _response;
+        private readonly object _lock = new object();
+
+        private TaskCompletionSource<bool> _taskCompletion;
+        private TextWriter _writer;
+        private JsonWriter _jsonWriter;
+        private bool _disposed;
+
+        public StdioResponseStream(ISharedTextWriter writer, ResponsePacket response)
+        {
+            _sharedWriter = writer;
+            _response = response;
+        }
+
+        public override bool CanWrite
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            CheckDisposed();
+
+            lock (_lock)
+            {
+                if (_taskCompletion == null)
+                {
+                    _taskCompletion = new TaskCompletionSource<bool>();
+                    _sharedWriter.Use(writer =>
+                    {
+                        try
+                        {
+                            _writer = writer;
+                            _jsonWriter = new JsonTextWriter(writer);
+
+                            _jsonWriter.WriteStartObject();
+                            _writer.Write(@"""body"":"); // WritePropertyName would mess up the JsonWriter state
+                            _writer.Write(Encoding.UTF8.GetChars(buffer, offset, count));
+
+                        }
+                        catch (Exception e)
+                        {
+                            _taskCompletion.SetException(e);
+                        }
+
+                        // we use a task to tell the shared text writer when we are
+                        // done using it
+                        return _taskCompletion.Task;
+                    });
+                }
+                else
+                {
+                    try
+                    {
+                        _writer.Write(Encoding.UTF8.GetChars(buffer, offset, count));
+                    }
+                    catch (Exception e)
+                    {
+                        _taskCompletion.SetException(e);
+                    }
+                }
+            }
+        }
+
+        public override void Flush()
+        {
+            // noop
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            lock (_lock)
+            {
+                CheckDisposed();
+                _disposed = true;
+            }
+
+            try
+            {
+                _writer.Write(",");
```

Comment:
do you mean the ```,``` or all of it (the rest of the object)? With the comma there is the case that when nothing has been written it's not needed